src/backend/web-gl/function-node.js

```diff
@@ -642,17 +644,78 @@ module.exports = class WebGLFunctionNode extends FunctionNodeBase {
 			}
 			const retDeclaration = [];
 			this.astGeneric(declaration, retDeclaration);
-			if (retDeclaration[2] === 'getImage2D(' || retDeclaration[2] === 'getImage3D(') {
-				if (i === 0) {
-					retArr.push('vec4 ');
+			let declarationType = 'Number';
+			if (i === 0) {
+				const init = declaration.init;
+				if (init) {
+					if (init.object) {
+						if (
+							init.object.type === 'MemberExpression' &&
+							init.object.object
+						) {
+							if (init.object.object.type === 'Identifier') {
+								const type = this.getParamType(init.object.object.name);
+								declarationType = typeLookupMap[type];
+								if (!declarationType) {
+									throw new Error(`unknown lookup type ${ typeLookupMap }`);
+								}
+								debugger;
+							} else if (
+								init.object.object.object &&
+								init.object.object.object.type === 'Identifier'
+							) {
+								const type = this.getParamType(init.object.object.object.name);
+								declarationType = typeLookupMap[type];
+								if (!declarationType) {
+									throw new Error(`unknown lookup type ${ typeLookupMap }`);
+								}
+								debugger;
+							} else if (
+								init.object.object.object.object
+							) {
+								if (
+									init.object.object.object.object.type === 'ThisExpression' &&
+									init.object.object.object.property.name === 'constants'
+								) {
+									const type = this.getConstantType(init.object.object.property.name);
+									declarationType = typeLookupMap[type];
+									if (!declarationType) {
+										throw new Error(`unknown lookup type ${ typeLookupMap }`);
+									}
+									debugger;
+								} else if (
+									init.object.object.object.object.object.type === 'ThisExpression' &&
```

Comment:
init.object.object.object.object.object.type -> this is also very sacry :P

Reply:
Haha, yea. In most of the cases is usually a quick work around AST nesting....

So i might be completely wrong (out of touch), for example `var foo = function foo(a) { ... }`

Because we cant do an AST scan using `function foo(a) {...}` itself, we did a hack to wrap it around into a `var foo`.

We then do an AST scan on that, so if we want to get the function content, it would be.

var -> foo -> assignment -> function -> function body, or something like that.

Hence the `object.object.object.object` because the `var foo = ` part is static which we defined.